### PR TITLE
ENH: datetime unpacker helpers

### LINKED ISSUE #13779: year, month, day convenience properties on datetime64
It would be really nice if numpy's datetime64's had convenience properties for `year`, `month`, `day` (of month), `hour`, `minute`, `second`, each of which would return an integer value. `dayofweek` and `dayofyear` etc would also be nice.

Pandas' DatetimeIndex does this, and it's really useful, but it's not great to have to convert data just to use that.

### PR DESCRIPTION
**Pull request by jklymak**
Closes #13779

This implements simple recipes in python for unpacking years, months, days, etc from `datetime64` arrays.  i.e.

```
    >>> d = np.arange('2015-12-20', 31*3, 15, dtype='datetime64[D]')
    >>> d
    array(['2015-12-20', '2016-01-04', '2016-01-19', '2016-02-03',
           '2016-02-18', '2016-03-04', '2016-03-19'], dtype='datetime64[D]')
    >>> m = np.datetime_month(d); m
    array([12,  1,  1,  2,  2,  3,  3])

```

Not sure about the performance of these functions versus going down to the c-level, but I think this is certainly better than many of the workarounds kicking around on stack overflow.  

- [ ] Squash commits

### CODE REVIEW

**Comment by jklymak** [comment]
Phew, wasn't readily apparent how to get the module name test to pass...`test_public_api.py`.  Not 100% sure I should have these python-only files in `multiarray.py`, but it works.

**Comment by eric-wieser** [comment]
I wonder whether we want a new submodule for this type of helper, maybe np.datetime

**Comment by jklymak** [comment]
like in `numpy/core/datetime.py`?  Happy to do that, but maybe someone more experienced w/ your code organization would be better after the PR?

**Comment by jklymak** [comment]
Does numpy build the docs as part of CI, and is the built version available after the build?  I can't get them to build locally.  

```
There is a programmable error in your configuration file:

Traceback (most recent call last):
  File "/Users/jklymak/anaconda3/envs/numpydev/lib/python3.7/site-packages/sphinx/config.py", line 361, in eval_config_file
    execfile_(filename, namespace)
  File "/Users/jklymak/anaconda3/envs/numpydev/lib/python3.7/site-packages/sphinx/util/pycompat.py", line 86, in execfile_
    exec(code, _globals)
  File "/Users/jklymak/numpy/doc/source/conf.py", line 372, in <module>
    class NumPyLexer(CLexer):
  File "/Users/jklymak/numpy/doc/source/conf.py", line 375, in NumPyLexer
    tokens = copy.deepcopy(lexers['c'].tokens)
KeyError: 'c'
```

**Comment by rgommers** [comment]
It does, and yes that failure is annoying. Sphinx 1.8.5 should work, the Sphinx 2.0 release broke something

**Comment by jklymak** [comment]
Ooops, sorry, didn't see that 1.8.5 was required and just did `conda install sphinx`.  

But... is the CI-built html available after running CI?

**Comment by seberg** [comment]
@jklymak it is here: https://8746-908607-gh.circle-artifacts.com/0/home/circleci/repo/doc/build/html/index.html unfortunately it is hidden away unless you are part of numpy, it would be nice to put that somewhere discoverable...

**Comment by WarrenWeckesser** [comment]
Certainly not discoverable, but you can see the CircleCI result with this pattern:

    https://circle-artifacts.com/gh/numpy/numpy/8746/artifacts/0/home/circleci/repo/doc/build/html/index.html 

In general, replace `8746` with the CircleCI job number.  I copied that idea from a similar pattern that was pointed out in a SciPy pull request.  I don't know if the above URL should be considered *the* canonical URL to use for a specific job, but it worked for me.  CircleCI redirected it to the URL that @seberg gave above.  There might be a simpler URL that would also work.

**Comment by seberg** [comment]
Ah, that is nicer. I doubt that there is more currently, there is also: https://circleci.com/api/v1.1/project/github/numpy/numpy/latest/artifacts?branch=pull/14276 but while that directly uses the PR number, it will only list the direct urls to all files for scripting use...

**Comment by jklymak** [comment]
OK, great!  https://circle-artifacts.com/gh/numpy/numpy/8747/artifacts/0/home/circleci/repo/doc/build/html/index.html 
is the latest version.

**Comment by charris** [comment]
Is `numpy/core/multiarray.py` where we want these? That file was added for backwards compatibility after the ufunc and multiarray merge. @eric-wieser 's idea for a new module sounds like a better solution, but you will need to explicitly import it into the `__init__.py` file.

**Comment by jklymak** [comment]
Sure.  But where do you want it?  Called: datetime_utils.py ?   It’s a little confusing to have these python only files here and the other ones in multiarray.

**Comment by jklymak** [comment]
Put into `numpy/core/datetimeutils.py`.  Not sure I got all the imports correct, so any guidance welcome.

**Comment by jklymak** [comment]
I think the CI error is spurious.

**Comment by jklymak** [comment]
- added release notes
- rebased

**Comment by mattip** [comment]
Rather than adding more functions to the top-level numpy namespace, I think these should become property getters on the datetime class instance. Unfortunately numpy will only support this after the dtype refactor.

**Comment by eric-wieser** [comment]
Another option that we can do today would be to make these class methods on `np.datetime64`

**Comment by jklymak** [comment]
I’m fine with whatever organization you see fit.  OTOH making these class methods will not be very discoverable.  Making these getters seems good.  I’m not, however, at all clear how either would be implemented given that the class is defined in c code.

**Comment by jklymak** [comment]
Cleaning up some old PRs.  Was there any more guidance on this?  I'm loathe to define these in the c code, given my general unfamiliarity and because I think they would be a lot harder to understand.  Is there a way to add them to the class, but implement in python?

**Comment by eric-wieser** [comment]
> Is there a way to add them to the class, but implement in python?

Yes, this is how `np.core._internal` is used - there are handfuls of c functions whose body is solely a `npy_cache_import` followed by a call back into a python implementation.

**Comment by jklymak** [comment]
Thanks @eric-wieser, can you give me an example, just so I can grep my way through and see how its done?

**Comment by eric-wieser** [comment]
Sure, here's an example:

https://github.com/numpy/numpy/blob/593ef5fc5a02fbcd6eeb70a59684b3b21c9cc643/numpy/core/src/multiarray/descriptor.c#L1910-L1923

You'll still need to work out how to define a static python method in C, but most of the body of the method can exist in python.

File: numpy/core/datetimeutils.py [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+
+    Examples
+    --------
+    >>> d = np.arange('2015-12-31', 12*3, 12, dtype='datetime64[M]')
+    >>> d
+    array(['2015-12', '2016-12', '2017-12'], dtype='datetime64[M]')
+    >>> y = np.datetime_year(d); y
+    array([2015, 2016, 2017])
+    """
+    return arr.astype('datetime64[Y]').astype(int) + 1970
```
Comment (eric-wieser) [inline numpy/core/datetimeutils.py:40]:
```suggestion
    return arr.astype('datetime64[Y]').astype(np.int64) + 1970
```

**Comment by jklymak** [comment]
I'm going to close this.  I think this would be useful, but I don't have the bandwidth to figure out passing references back and forth between C and python.  Of course anyone should feel free to pick this up if there is interest!

**Pull request closed without merging**